File: cpp/cpp_100-999/655.cc
// C++
// leetcode 655
// https://leetcode.cn/problems/print-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<string>> printTree(TreeNode* root) {
    	int depth = 0, m = 1;
    	queue<TreeNode*> q;
    	q.push(root);
    	TreeNode* tmp;

    	// find depth
    	while (!q.empty()) {
    		int size = q.size();
    		m <<= 1;
    		depth++;
    		while(size--) {
    			tmp = q.front();
    			q.pop();
    			if (tmp -> left)
    				q.push(tmp -> left);
    			if (tmp -> right)
    				q.push(tmp -> right);
    		}
    	}

    	vector<vector<string>> res(depth, vector<string>(m - 1, ""));
    	queue<pair<TreeNode*, int>> qq;


    	qq.push(make_pair(root, (m - 2) / 2));
    	int level = 0;
    	int idx = -1, d = m>>2;
    	while (!qq.empty()) {
    		int size = qq.size();
    		while (size--) {
    			tmp = qq.front().first;
    			idx = qq.front().second;
    			res[level][idx] = to_string(tmp -> val);
    			qq.pop();
    			if (tmp -> left)
    				qq.push(make_pair(tmp -> left, idx - d));
    			if (tmp -> right)
    				qq.push(make_pair(tmp -> right, idx + d));
    		}
    		level++;
    		d >>= 1;
    	}

    	return res;
    }
};

File: Go/Go_100-999/655.go
// Go
// leetcode 655
// https://leetcode.cn/problems/print-binary-tree/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import (
	"strconv"
)

func printTree(root *TreeNode) [][]string {
	depth, m := 0, 1
	q_p := []*TreeNode{root}

	// find depth
	for len(q_p) != 0 {
		depth++
		size := len(q_p)
		for i := 0; i < size; i++ {
			if q_p[i].Left != nil {
				q_p = append(q_p, q_p[i].Left)
			}
			if q_p[i].Right != nil {
				q_p = append(q_p, q_p[i].Right)
			}
		}
		q_p = q_p[size : ]
	}
	m <<= depth

	// preteatment for res
	res := make([][]string, depth)
	for i := 0; i < depth; i++ {
		res[i] = make([]string, m - 1)
	}

	// fill res
	q_p = append(q_p, root)
	q_idx := []int{(m - 2) / 2}
	level, d := 0, m >> 2
	for len(q_p) != 0 {
		size := len(q_p)

		for i := 0; i < size; i++ {
			res[level][q_idx[i]] = strconv.Itoa(q_p[i].Val)
			if q_p[i].Left != nil {
				q_p = append(q_p, q_p[i].Left)
				q_idx = append(q_idx, q_idx[i] - d)
			}
			if q_p[i].Right != nil {
				q_p = append(q_p, q_p[i].Right)
				q_idx = append(q_idx, q_idx[i] + d)
			}
		}
		level++
		d >>= 1
		q_p = q_p[size : ]
		q_idx = q_idx[size : ]
	}

	return res
}